### DISCUSSION TITLE
Extracting CRPS from a Random Survial Forest

### DISCUSSION THREAD

**Discussion by setseed13** [kogalur/randomForestSRC#400] [Q&A] [answered]
Hi!

Right now, after having trained a couple of RSF, I try to extract some statistics from the RSF objects, and I struggle to get the CRPS.

As per your documentation to draw the CRPS over time, we can use the following:
```
library(randomForestSRC)
data(veteran, package = "randomForestSRC")

v.obj <- rfsrc(Surv(time, status) ~ .,
    data = veteran,
    ntree = 100, block.size = 1
)

bs.rsf <- get.brier.survival(v.obj, cens.model = "rfsrc")$brier.score

trapz <- randomForestSRC:::trapz
time <- v.obj$time.interest
crps.rsf <- sapply(1:length(time), function(j) {
    trapz(time[1:j], bs.rsf[1:j, 2] / diff(range(time[1:j])))
})

plot(time, crps.rsf, ylab = "CRPS", type = "s", col = 2)
```

Hence, I thought that I could calculate the CRPS via

```
trapz(time[1:length(time)], bs.rsf[1:length(time), 2] / diff(range(time[1:length(time)])))
```

However, it differs from the one shown via

```
v.obj
```

Is there a problem with the intervall I passed into `trapz`?

Additionally, I looked into the source of `print.rsf.R` and found
```
crps.err  <- get.brier.survival(x)$crps.std
```

So, I tried
```
get.brier.survival(v.obj, cens.model = "rfsrc")$crps.std
```

and get yet another value. I'd love to get some pointers at why these three approaches differ, and which of these is the "true" CRPS.

Thanks!

**Comment by ishwaran** [comment, marked as answer]
The problem has to do with Monte Carlo variability.  You are using the option `cens.model = "rfsrc"` which has the effect that the censoring distribution is estimated by running a random survival forest analysis.  Such an analysis by definition is random and therefore the forest estimated censoring distribution will always change each time (although slightly) which will cause CRPS to change each time a call is made.

Also I would add that `get.brier.survival` uses the default option 'cens.model = "km"' and therefore the CRPS output obtained by printing the object `v.obj` uses the Kaplan-Meier estimator for the censoring distribution which is different than using random survival forests.   Note that this CRPS value will not change as there is no MC variability in its calculation.

**Reply by setseed13** [reply]
Thank you so much for the reply.

I thought, since the CRPS is defined via an integral, it would be calculated purely analytically from the (stored) Brier scores, and didn't suspect probabilistic methods to be used each call. 

After reading your reply I looked at the code again. `cens.model = rfsrc` then hinted that the Brier scores are not actually stored in `v.obj`, and that it rather just holds the information on how to build the forest.

Running then the following code confirmed that.
```
v.obj_1 <- rfsrc(Surv(time, status) ~ .,
    data = veteran,
    ntree = 100, block.size = 1
)
v.obj_2 <- rfsrc(Surv(time, status) ~ .,
    data = veteran,
    ntree = 100, block.size = 1
)
v.obj_1
v.obj_2
```

So, thans, I now understand what's going on!
